### PR TITLE
Add basic support for OpenBSD

### PR DESCRIPTION
**Pull request by messense**
Closes #469 

cc @octeep

### CODE REVIEW

File: src/module_writer.rs [hunk 1]
```diff
@@ -394,7 +394,7 @@ fn cffi_init_file() -> &'static str {
 import os
 from .ffi import ffi
 
-lib = ffi.dlopen(os.path.join(os.path.dirname(__file__), 'native.so'), 4098)
```
Comment (messense) [inline src/module_writer.rs:397]: Why was this `4098`? It raises `Invalid mode` on OpenBSD, it does not have `RTLD_NODELETE`.

**Comment by messense** [comment]
I've verified that `cargo test` passed with this branch on OpenBSD 6.8.

**Comment by dkarama** [comment]
I'm facing this issue https://github.com/ArniDagur/python-adblock/issues/36. I though it would be cleared with your patches though it still persists. What do you think is causing this?

On `OpenBSD-current` with rust-1.51.0.

[test.log](https://github.com/PyO3/maturin/files/6291692/test.log)

**Comment by messense** [comment]
> I'm facing this issue [ArniDagur/python-adblock#36](https://github.com/ArniDagur/python-adblock/issues/36). I though it would be cleared with your patches though it still persists. What do you think is causing this?
> 
> On `OpenBSD-current` with rust-1.51.0.
> 
> [test.log](https://github.com/PyO3/maturin/files/6291692/test.log)

I'm not sure, can you list down the steps to reproduce the issue?

**Comment by dkarama** [comment]
> > I'm facing this issue [ArniDagur/python-adblock#36](https://github.com/ArniDagur/python-adblock/issues/36). I though it would be cleared with your patches though it still persists. What do you think is causing this?
> > On `OpenBSD-current` with rust-1.51.0.
> > [test.log](https://github.com/PyO3/maturin/files/6291692/test.log)
> 
> I'm not sure, can you list the steps to reproduce the issue?

The Makefile for this in ports is attached, only thing worth looking there in are the crates used while compiling.

[Makefile.txt](https://github.com/PyO3/maturin/files/6291857/Makefile.txt)

And the build command is:
`/usr/local/bin/cargo build  --manifest-path /usr/ports/pobj/maturin-0.10.2.20200411/maturin-bdf4bb23a12abec2f944167546966b057744d3eb/Cargo.toml  --offline  --release  --verbose   ;`

But if you have a OpenBSD machine and know how to use ports, try [jasperla/openbsd-wip](https://github.com/jasperla/openbsd-wip), I've recently commited the ports as devel/maturin and www/py-adblock.

**Comment by messense** [comment]
I've reproduce the issue with just `maturin build`, but I don't know how to install rust nightly on OpenBSD thus I'm unable to use [cargo-expand](https://github.com/dtolnay/cargo-expand) to debug the `pyo3-macros` proc-macro generated code.

**Comment by dkarama** [comment]
Are you using the ports system? Or downloading the binaries? There is no hope with binary packages. But with ports edit Makefile in lang/rust to have `DISTNAME = rustc-nightly-src`. Then `make makesum` and make install. Nightly versions have never been tested and I'm not sure if the build will be successful. I'm trying to build right now.

**Comment by messense** [comment]
@dkarama I was able to get tests pass with pyo3 git main branch code. Try  patch python-adblock's `Cargo.toml`:

```toml
[patch.crates-io]
pyo3 = { git = "https://github.com/PyO3/pyo3.git", rev = "7ac75366c27d5be6cc8fd5b344da1953c12d2c3e" }
```

**Comment by dkarama** [comment]
@messense 

Did you patch something else? The latest HEAD on python-adblock fails with to compile your changes to `Cargo.toml`

```
Building a mixed python/rust project
 Found pyo3 bindings
 Found CPython 3.8 at python3.8
 Built source distribution to /home/zen/Github/python-adblock/target/wheels/adblock-0.4.3.tar.gz
   Compiling adblock v0.4.3 (/home/zen/Github/python-adblock)
error: hidden lifetime parameters in types are deprecated
   --> src/lib.rs:206:52
    |
206 |     pub fn add_filter_list(&mut self, filter_list: &str, format: &str) -> PyResult<()> {
    |                                                    ^- help: indicate the anonymous lifetime: `<'_>`
    |
note: the lint level is defined here
   --> src/lib.rs:5:5
    |
5   |     rust_2018_idioms,
    |     ^^^^^^^^^^^^^^^^
    = note: `#[deny(elided_lifetimes_in_paths)]` implied by `#[deny(rust_2018_idioms)]`


.........

error: aborting due to 20 previous errors


error: could not compile `adblock`

To learn more, run the command again with --verbose.
 maturin failed
  Caused by: Failed to build a native library through cargo
  Caused by: Cargo build finished with "exit code: 101": `cargo rustc --message-format json --manifest-path Cargo.toml --release --lib --`
```

**Comment by messense** [comment]
Just remove the rust_2018_idioms lint.

File: src/module_writer.rs (hunk 1 above)
Comment (konstin) [reply src/module_writer.rs:397]: No idea tbh. I thought I copied that code from an example, but I can't find that anymore :eyes:.

**Review by konstin — APPROVED** [review]
Thanks, looks good

**Pull request merged**